packages/astro/src/vite-plugin-mdx/index.ts

```diff
@@ -0,0 +1,134 @@
+import type { TransformResult } from 'rollup';
+import {
+	transformWithEsbuild,
+	type Plugin,
+	type ResolvedConfig,
+} from 'vite';
+import type { AstroRenderer, AstroSettings } from '../@types/astro';
+import type { LogOptions } from '../core/logger/core.js';
+import type { PluginMetadata } from '../vite-plugin-astro/types';
+
+import babel from '@babel/core';
+import { CONTENT_FLAG, PROPAGATED_ASSET_FLAG } from '../content/index.js';
+import { astroEntryPrefix } from '../core/build/plugins/plugin-component-entry.js';
+import { removeQueryString } from '../core/path.js';
+import tagExportsPlugin from './tag.js';
+
+interface TransformJSXOptions {
+	code: string;
+	id: string;
+	mode: string;
+	renderer: AstroRenderer;
+	ssr: boolean;
+	root: URL;
+}
+
+async function transformJSX({
+	code,
+	mode,
+	id,
+	ssr,
+	renderer,
+	root,
+}: TransformJSXOptions): Promise<TransformResult> {
+	const { jsxTransformOptions } = renderer;
+	const options = await jsxTransformOptions!({ mode, ssr });
+	const plugins = [...(options.plugins || [])];
+	if (ssr) {
+		plugins.push(await tagExportsPlugin({ rendererName: renderer.name, root }));
+	}
+	const result = await babel.transformAsync(code, {
+		presets: options.presets,
+		plugins,
+		cwd: process.cwd(),
+		filename: id,
+		ast: false,
+		compact: false,
+		sourceMaps: true,
+		configFile: false,
+		babelrc: false,
+		inputSourceMap: options.inputSourceMap,
+	});
+	// TODO: Be more strict about bad return values here.
+	// Should we throw an error instead? Should we never return `{code: ""}`?
+	if (!result) return null;
+
+	if (renderer.name === 'astro:jsx') {
+		const { astro } = result.metadata as unknown as PluginMetadata;
+		return {
+			code: result.code || '',
+			map: result.map,
+			meta: {
+				astro,
+				vite: {
+					// Setting this vite metadata to `ts` causes Vite to resolve .js
+					// extensions to .ts files.
+					lang: 'ts',
+				},
+			},
+		};
+	}
+
+	return {
+		code: result.code || '',
+		map: result.map,
+	};
+}
+
+interface AstroPluginJSXOptions {
+	settings: AstroSettings;
+	logging: LogOptions;
+}
+
+// Format inspired by https://github.com/vitejs/vite/blob/main/packages/vite/src/node/constants.ts#L54
+const SPECIAL_QUERY_REGEX = new RegExp(
+	`[?&](?:worker|sharedworker|raw|url|${CONTENT_FLAG}|${PROPAGATED_ASSET_FLAG})\\b`
+);
+
+/** Use Astro config to allow for alternate or multiple JSX renderers (by default Vite will assume React) */
+export default function mdxVitePlugin({ settings }: AstroPluginJSXOptions): Plugin {
```

Comment:
In the future, could we move this MDX Vite plugin to `@astrojs/mdx`?

Reply:
Yeah, I can try that. I'm not sure why this separation exists.

Reply:
Cool, we can also do this as a follow-up if you prefer 😄